zerver/lib/bugdown/api_code_examples.py

```diff
@@ -77,11 +87,93 @@ def render_python_code_example(function: str, admin_config: Optional[bool]=False
+    valid_methods = ["GET", "POST", "DELETE", "PUT", "PATCH", "OPTIONS"]
+    if method == valid_methods[0]:
+        # Then we need to make sure that each -d option translates to becoming
+        # a GET parameter (in the URL) and not a POST parameter (in the body).
+        # TODO: remove the -X part by updating the linting rule. It's redundant.
+        return ["-X", "GET", "-G", url, "\\"]
+    elif method in valid_methods:
+        return ["-X", method, url, "\\"]
+    else:
+        msg = "The request method {} is not one of {}".format(method,
+                                                              valid_methods)
+        raise ValueError(msg)
+
+def generate_curl_example(endpoint: str, method: str,
+                          auth_email: str=DEFAULT_AUTH_EMAIL,
+                          auth_api_key: str=DEFAULT_AUTH_API_KEY,
+                          api_url: str=DEFAULT_API_URL) -> List[str]:
+    lines = ["```curl"]
+    openapi_entry = openapi_spec.spec()['paths'][endpoint][method.lower()]
+
+    curl_first_line_parts = ["curl"] + curl_method_arguments(endpoint, method,
+                                                             api_url)
+    lines.append(" ".join(curl_first_line_parts))
+
+    authentication_required = openapi_entry.get("security", False)
+    if authentication_required:
+        lines.append("    -u %s:%s \\" % (auth_email, auth_api_key))
+
+    openapi_example_params = get_openapi_parameters(endpoint, method)
+    for packet in openapi_example_params:
+        param_name = packet["name"]
+        param_type = packet["schema"]["type"]
+        if param_type in ["object", "array"]:
+            example_value = packet.get("example", None)
+            if not example_value:
+                msg = """All array and object type request parameters must have
+concrete examples. The openAPI documentation for {}/{} is missing an example
+value for the {} parameter. Without this we cannot automatically generate a
+cURL example.""".format(endpoint, method, param_name)
+                raise ValueError(msg)
+            if type(example_value) == dict:
+                ordered_example_value = {}
+                for key in sorted(example_value.keys()):
+                    ordered_example_value[key] = example_value[key]
+            elif type(example_value) == list:
+                ordered_example_value = []  # type: ignore # there is no issue with this line.
+                for element in sorted(example_value):
+                    ordered_example_value.append(element)  # type: ignore # it's a list, not a dict.
+            ordered_ex_val_str = str(ordered_example_value).replace("'", '"')
+            line = "    --data-urlencode {}='{}' \\".format(param_name, ordered_ex_val_str)
+        else:
+            example_value = packet.get("example", DEFAULT_EXAMPLE[param_type])
+            line = "    -d '{}={}' \\".format(param_name, example_value)
+        lines.append(line)
+
+    lines[-1] = lines[-1][:-2]
```

Comment:
This is to stripe off the `\\`s, right?  I think it'd be better to have a loop here that adds `\\`s to all but the last line.

Reply:
(It'd make the above logic a lot simpler)

Reply:
I actually think that that would complicate it and messy up the code a bit too. But I've added an inline comment now to make things clearer. Thanks for bringing it to my attention.

Reply:
I'm not sure I understand why that would be more complex -- we'd have `\\` appear in just one place rather than 4.

Reply:
I was thinking that replacing just the one line which did the `\\` stripping at the end would be less complex/messy then having a second loop iterate through all of the lines (except the last one) adding `\\` to the end. Though, I see your point too, so I'll just go ahead and make the change.